manager/db.go

```diff
@@ -46,49 +49,144 @@ func (m *Manager) lookupCatalogs(environmentId string) ([]model.Catalog, error)
 	return catalogs, nil
 }
 
-func (m *Manager) lookupCatalog(environmentId, name string) (model.Catalog, error) {
+func (m *Manager) lookupCatalog(environmentId, name string) (model.CatalogModel, error) {
 	var catalogModel model.CatalogModel
 	if err := m.db.Where(&model.CatalogModel{
 		Catalog: model.Catalog{
 			EnvironmentId: environmentId,
 			Name:          name,
 		},
 	}).First(&catalogModel).Error; err != nil {
-		return model.Catalog{}, err
+		return model.CatalogModel{}, err
 	}
-	return catalogModel.Catalog, nil
+	return catalogModel, nil
 }
 
-func (m *Manager) updateDb(catalog model.Catalog, templates []model.Template, newCommit string) error {
-	tx := m.db.Begin()
+func (m *Manager) deleteTemplates(templates []model.TemplateModel, templateCommits, templateLookup map[string]bool, tx *gorm.DB) error {
+	for _, dbTemplate := range templates {
+		// template either changed or is deleted
+		if _, ok := templateCommits[dbTemplate.Commit]; !ok {
+			if _, ok := templateLookup[dbTemplate.FolderName]; !ok {
+				// delete from db only when template is deleted
+				if err := tx.Where(&dbTemplate).Delete(&model.TemplateModel{}).Error; err != nil {
+					tx.Rollback()
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
 
-	if err := tx.Where(&model.CatalogModel{
-		Catalog: model.Catalog{
-			Name:          catalog.Name,
-			EnvironmentId: catalog.EnvironmentId,
+func (m *Manager) lookupTemplates(catalogId uint, environmentId string) ([]model.TemplateModel, error) {
+	var templates []model.TemplateModel
+	if err := m.db.Where(&model.TemplateModel{
+		Template: model.Template{
+			CatalogId:     catalogId,
+			EnvironmentId: environmentId,
 		},
-	}).Delete(&model.CatalogModel{}).Error; err != nil {
-		tx.Rollback()
-		return err
+	}).Find(&templates).Error; err != nil {
+		return nil, err
 	}
+	return templates, nil
+}
 
-	var catalogModel model.CatalogModel
+func (m *Manager) lookupVersions(templateIDs []uint) ([]model.VersionModel, error) {
+	var versions []model.VersionModel
+	versionQuery := `
+	SELECT *
+	FROM catalog_version
+	WHERE template_id IN (?)`
+	if err := m.db.Raw(versionQuery, templateIDs).Find(&versions).Error; err != nil {
+		return nil, err
+	}
+	return versions, nil
+}
+
+func (m *Manager) saveCatalog(dbCatalogId uint, catalog model.Catalog, commit string, tx *gorm.DB) (model.CatalogModel, error) {
```

Comment:
It might be clearer if this function was split into two. It's two code paths that are pretty disjoint for the most part. Maybe `updateCatalog` and `createCatalog`?

Reply:
I can break it down further within save catalog to contain those functions. But I would like to keep saveCatalog because I think it is descriptive of what I want to happen. If it doesnt exist, create it. If it does exist and a change was made, update it.